torch/fx/experimental/Partitioner.py

```diff
@@ -0,0 +1,145 @@
+        line += 'all nodes in this partition: '
+        for node in self.nodes:
+            line += '(' + node.name + ':' + node.op + ') '
+        line += '\n'
+        line += 'parent partitions:'
+        for parent in self.parents:
+            if parent.partition_id == 0:
+                line += 'root_partition '
+            else:
+                line += 'partition_' + str(parent.partition_id) + ' '
+        line += '\n'
+        line += 'children partitions:'
+        for child in self.children:
+            line +=  str(child.partition_id) + ' '
+        line += '\n'
+        line += 'input nodes: '
+        input_nodes = self.get_input_nodes()
+        for node in input_nodes:
+            line += '(' + node.name + ':' + node.op + ') '
+        line += '\n'
+        line += 'output nodes: '
+        output_nodes = self.get_output_nodes()
+        for node in output_nodes:
+            line += '(' + node.name + ':' + node.op + ') '
+        return line
+
+class Partitioner:
+    def __init__(self) -> None:
+        """
+        After a partitioner is created,
+        it first check if multiple types of devices (backends) are involved.
+        So far, we assume that there is only one backend and one device with unlimited memory
+        """
+        self.partitions = []
+        self.devices = []
+        self.graph_module = None
+        self.node_to_partitions = {}
+
+    def partition_graph(self, fx_module: GraphModule, devices: List[str]) -> Dict[Partition, List[Partition]]:
+        """
+        Given the fx_moudle and devices, find the partitions, do the partitions,
+        and then return a dictionary representing the DAG of partitions
+        """
+        self.graph_module = fx_module
+        self.devices = devices
+        #Create a dummy partition as root.
+        self.root_partition = self.create_partition()
+        if len(self.devices) == 1:
+            self.find_single_partition()
+        #TODO: size_based_partition for mutiple devices
+        else:
+            pass
+        self.do_partition()
+        dag_graph = self.dump_partition_DAG()
+        return dag_graph
+
+    def find_single_partition(self) -> None:
+        """Only one partition (one graph on one device)."""
+        partition_0 = self.create_partition()
+        for i, node in enumerate(self.graph_module.graph.nodes):
+            self.node_to_partitions[node] = [partition_0.partition_id]
+            partition_0.add_node(node)
+        #Connect the partition to the root.
+        self.root_partition.add_child(partition_0)
+        partition_0.add_parent(self.root_partition)
+        return
+
+    def do_partition(self) -> None:
+        """Mark the partition on each node in the fx_module."""
+        for node in self.graph_module.graph.nodes:
+            node.partition_ids = self.node_to_partitions[node]
+        return
+
+    def dump_partition_DAG(self) -> Dict[Partition, List[Partition]]:
```

Comment:
I think I'd prefer if the DAG was a static structure and not a dict of Partitions. Basically all the information should be calculated here and returned in the DAG. 
Inputs/Outputs should be a static list, same for sizes eventually. This way the information is preserved should anything happen to the graph itself.